mev_inspect/compound_liquidations.py

```diff
@@ -0,0 +1,113 @@
+    raise RuntimeError("Missing environment variable RPC_URL")
+else:
+    w3 = Web3(Web3.HTTPProvider(rpc))
+
+V2_COMPTROLLER_ADDRESS = "0x3d9819210A31b4961b30EF54bE2aeD79B9c9Cd3B"
+V2_C_ETHER = "0x4Ddc2D193948926D02f9B1fE9e1daa0718270ED5"
+WETH_ADDRESS = "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2"
+
+# cToken=>Token mapping (cDAI=>DAI)
+# useful for finding out the underlying asset address of the cToken seized
+def get_all_comp_markets():
+    c_token_mapping = {}
+    comp_v2_comptroller_abi = get_raw_abi("Comptroller", Protocol.compound_v2)
+    comptroller_instance = w3.eth.contract(
+        address=V2_COMPTROLLER_ADDRESS, abi=comp_v2_comptroller_abi
+    )
+    markets = comptroller_instance.functions.getAllMarkets().call()
+    for c_token in markets:
+        # make an exception for cETH (as it has no .underlying())
+        if c_token != V2_C_ETHER:
+            comp_v2_ctoken_abi = get_raw_abi("CToken", Protocol.compound_v2)
+            ctoken_instance = w3.eth.contract(address=c_token, abi=comp_v2_ctoken_abi)
+            underlying_token = ctoken_instance.functions.underlying().call()
+            c_token_mapping[
+                c_token.lower()
+            ] = underlying_token.lower()  # make k:v lowercase for consistancy
+    return c_token_mapping
+
+
+# # find if the searcher repays the loan from their own EOA, by buying it from a DEX, or w/ a flashloan
+# def find_collateral_source(
+#     classified_traces: list[ClassifiedTrace],
+#     tx: Transaction,
+#     liquidation_contract: Optional[str],
+# ) -> LiquidationCollateralSource:
+#     source = LiquidationCollateralSource.other  # set other by default
+#     for classified_trace in classified_traces:
+#         # look for trace that liquidates and see from address
+#         if (
+#             classified_trace.to_address == liquidation_contract
+#             and classified_trace.function_name == "liquidateBorrow"
+#         ):
+#             ## check if tx originates from searcher eoa or contract
+#             if tx.to_address.lower() == classified_trace.from_address:
+#                 source = LiquidationCollateralSource.searcher_contract
+#             elif tx.from_address.lower() == classified_trace.from_address:
+#                 source = LiquidationCollateralSource.searcher_eoa
+#             ## flashloan providers identified here
+#     return source
+
+
+def get_underlying_ctoken_exchange_rate(
+    c_token_address: str, block_number: int
+) -> float:
+    comp_v2_ctoken_abi = get_raw_abi("CToken", Protocol.compound_v2)
+    ctoken_instance = w3.eth.contract(address=c_token_address, abi=comp_v2_ctoken_abi)
+    raw_exchange_rate = ctoken_instance.functions.exchangeRateCurrent().call(
+        block_identifier=block_number
+    )
+    # format based on decimals in ctoken and the underlying token
+    # see "Interpreting Exchange Rates" https://compound.finance/docs#protocol-math
+    underlying_token_address = get_all_comp_markets()[c_token_address.lower()]
```

Comment:
similar comment here on passing markets instead of fetching to separate network calls from the logic operating on the data